cub/device/dispatch/dispatch_reduce.cuh

```diff
@@ -55,898 +55,1291 @@ CUB_NAMESPACE_BEGIN
+                                                      int ptx_version)
+      : d_temp_storage(d_temp_storage)
+      , temp_storage_bytes(temp_storage_bytes)
+      , d_in(d_in)
+      , d_out(d_out)
+      , num_items(num_items)
+      , reduction_op(reduction_op)
+      , init(init)
+      , stream(stream)
+      , ptx_version(ptx_version)
+  {} 
+
+  CUB_DETAIL_RUNTIME_DEBUG_SYNC_IS_NOT_SUPPORTED
+  CUB_RUNTIME_FUNCTION __forceinline__
+  DispatchReduce(void* d_temp_storage,
+                 size_t &temp_storage_bytes,
+                 InputIteratorT d_in,
+                 OutputIteratorT d_out,
+                 OffsetT num_items,
+                 ReductionOpT reduction_op,
+                 InitT init,
+                 cudaStream_t stream,
+                 bool debug_synchronous,
+                 int ptx_version)
+    : d_temp_storage(d_temp_storage)
+    , temp_storage_bytes(temp_storage_bytes)
+    , d_in(d_in)
+    , d_out(d_out)
+    , num_items(num_items)
+    , reduction_op(reduction_op)
+    , init(init)
+    , stream(stream)
+    , ptx_version(ptx_version)
+  {
+    CUB_DETAIL_RUNTIME_DEBUG_SYNC_USAGE_LOG
+  }
+
+  //---------------------------------------------------------------------------
+  // Small-problem (single tile) invocation
+  //---------------------------------------------------------------------------
+
+  /**
+   * @brief Invoke a single block block to reduce in-core
+   *
+   * @tparam ActivePolicyT
+   *   Umbrella policy active for the target device
+   *
+   * @tparam SingleTileKernelT
+   *   Function type of cub::DeviceReduceSingleTileKernel
+   *
+   * @param[in] single_tile_kernel
+   *   Kernel function pointer to parameterization of 
+   *   cub::DeviceReduceSingleTileKernel
+   */
+  template <typename ActivePolicyT, typename SingleTileKernelT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t
+  InvokeSingleTile(SingleTileKernelT single_tile_kernel)
+  {
+    cudaError error = cudaSuccess;
+    do
     {
-      CUB_DETAIL_RUNTIME_DEBUG_SYNC_USAGE_LOG
-    }
-
-
-    //------------------------------------------------------------------------------
-    // Small-problem (single tile) invocation
-    //------------------------------------------------------------------------------
-
-    /// Invoke a single block block to reduce in-core
-    template <
-        typename                ActivePolicyT,          ///< Umbrella policy active for the target device
-        typename                SingleTileKernelT>      ///< Function type of cub::DeviceReduceSingleTileKernel
-    CUB_RUNTIME_FUNCTION __forceinline__
-    cudaError_t InvokeSingleTile(
-        SingleTileKernelT       single_tile_kernel)     ///< [in] Kernel function pointer to parameterization of cub::DeviceReduceSingleTileKernel
+      // Return if the caller is simply requesting the size of the storage
+      // allocation
+      if (d_temp_storage == NULL)
+      {
+        temp_storage_bytes = 1;
+        break;
+      }
```

Comment:
This seems to be an invariant, so I would prefer if we could move it to the top of the function and turn the `break` into an explicit `return cudaSuccess;`

Reply:
I'd prefer this as well. I think it might be a good idea to have a cleanup PR that removes all `do{} while(0);` from the code.